public_html/wp-content/plugins/wordcamp-participation-notifier/wordcamp-participation-notifier.php

```diff
@@ -156,19 +150,128 @@ protected function published_speaker_post_updated( $post ) {
 	 * @param string  $old_status
 	 * @param WP_Post $post
 	 */
-	protected function speaker_post_published_or_unpublished( $new_status, $old_status, $post ) {
+	protected function post_published_or_unpublished( $new_status, $old_status, $post ) {
 		if ( 'publish' == $new_status ) {
-			$this->remote_post( self::PROFILES_HANDLER_URL, $this->get_post_activity_payload( $post ) );
-			$this->remote_post( self::PROFILES_HANDLER_URL, $this->get_post_association_payload( $post, 'add' ) );
-		} elseif( 'publish' == $old_status ) {
-			// Get the $user_id from post meta instead of $_POST in case it changed during the unpublish update.
-			// This makes sure that the association is removed from the same user that it was originally added to.
+			$user_id = $this->get_new_wporg_user_id( $post );
 
+			if ( ! get_user_meta( $user_id, $this->get_published_activity_key( $post ) ) ) {
+				$this->remote_post( self::PROFILES_HANDLER_URL, $this->get_post_activity_payload( $post ) );
+				update_user_meta( $user_id, $this->get_published_activity_key( $post ), true );
+			}
+
+			$this->add_badge( $post, $user_id );
+		} elseif( 'publish' == $old_status ) {
 			$user_id = $this->get_saved_wporg_user_id( $post );
+			$this->maybe_remove_badge( $post, $user_id );
+		}
+	}
+
+	/**
+	 * Makes request to Profile URL to add badge to organizer/speaker. Also adds a meta entry which is used by `maybe_remove_badge` function to figure out whether to remove a badge or not.
+	 *
+	 * @param WP_Post $post     Speaker/Organizer Post Object.
+	 * @param int     $user_id  User ID to add badge for.
+	 */
+	protected function add_badge( $post, $user_id ) {
+		if ( ! in_array( $post->post_type, [ 'wcb_speaker', 'wcb_organizer' ] ) || ! $user_id ) {
+			return;
+		}
+
+		$meta_key = $this->get_user_meta_key( $post );
+
+		// User already has a badge. Prevent wasteful API call and bail.
+		if ( get_user_meta( $user_id, $meta_key ) ) {
+			return;
+		}
+
+		update_user_meta( $user_id ,$meta_key, true );
+
+		$this->remote_post( self::PROFILES_HANDLER_URL, $this->get_post_association_payload( $post, 'add', $user_id ) );
+	}
+
+	/**
+	 * Makes a request to remove speaker/organizer badge from a user if the user is removed from all WordCamp where they were speaker/organizer.
+	 *
+	 * @param WP_Post $post     Speaker/Organizer Post Object.
+	 * @param int     $user_id  User ID to remove the badge for.
+	 */
+	protected function maybe_remove_badge( $post, $user_id ) {
+		global $wpdb;
+
+		if ( ! in_array( $post->post_type, [ 'wcb_speaker', 'wcb_organizer' ] ) ) {
+			return;
+		}
```

Comment:
Could the `is_post_notifiable` method be used here instead? That would make this DRYer and easier to refactor if we ever added another post type.